stratagem-runner/src/stratagem_purge.rs

```diff
@@ -4,11 +4,60 @@
 
 // Usage: stratagem_purge FS [NID1] [NID2] ...
 
-use std::env;
+use std::fs::File;
+use std::io;
+use std::io::prelude::*;
+use std::io::BufReader;
+use std::process::exit;
+use structopt::StructOpt;
+
+#[derive(Debug, StructOpt)]
+#[structopt(name = "stratagem_purge")]
+struct Opt {
+    #[structopt(short = "i")]
+    /// File to read from, "-" for stdin
+    input: Option<String>,
+
+    #[structopt(name = "FSNAME")]
+    /// Lustre filesystem name, or mountpoint
+    fsname: String,
+
+    #[structopt(name = "FIDS")]
+    /// Optional list of FIDs to purge
+    fidlist: Vec<String>,
+}
 
 fn main() {
-    let mut args = env::args();
-    let device = args.nth(1).expect("No device specified");
+    env_logger::init();
+    let opt = Opt::from_args();
+    let device = opt.fsname;
+    let input: Box<Iterator<Item = String>> = match opt.input {
+        None => {
+            if opt.fidlist.len() == 0 {
+                Box::new(BufReader::new(io::stdin()).lines().map(|x| x.unwrap()))
+            } else {
+                Box::new(opt.fidlist.iter().map(|x| x.to_owned()))
+            }
+        }
+        Some(name) => {
+            let buf: Box<BufRead> = match name.as_ref() {
+                "-" => Box::new(BufReader::new(io::stdin())),
```

Comment:
I get complicated return errors (which is why I ended up with Boxing everything.
If I move the code into a separate function:
```
  --> stratagem-runner/src/cli.rs:20:17
   |
14 |  /             if fidlist.is_empty() {
15 |  |                 BufReader::new(io::stdin())
   |  |_________________-
16 | ||                     .lines()
17 | ||                     .map(|x| x.expect("Failed to readline from stdin")
18 | ||                     )
   | ||_____________________- expected because of this
19 |  |             } else {
20 |  |                 fidlist.into_iter()
   |  |                 ^^^^^^^^^^^^^^^^^^^ expected struct `std::iter::Map`, found struct `std::vec::IntoIter`
21 |  |             }
   |  |_____________- if and else have incompatible types
   |
   = note: expected type `std::iter::Map<std::io::Lines<std::io::BufReader<std::io::Stdin>>, [closure@stratagem-runner/src/cli.rs:17:26: 17:71]>`
              found type `std::vec::IntoIter<std::string::String>`

error: aborting due to 2 previous errors
```

Reply:
bleh, yeah you need to unify the return types for all the branches. Futures has the concept of either to unify the types and avoid sticking things into `Box`:

https://docs.rs/futures/0.1.26/src/futures/future/either.rs.html#6-11

There is also an either crate (https://docs.rs/either/1.5.2/either/) that can be used for this (it implements `Iterator`) and can be used to avoid sticking each branch in a `Box`.

Reply:
Is it terrible to stick everything in a box?

Reply:
not terrible, but sticking something in a Box means allocating it on the heap instead of the stack from what I understand.